0000-settlement-engine/0000-settlement-engine.md

```diff
@@ -0,0 +1,353 @@
+
+Payment channels offer very fast and cheap settlement, which is beneficial to peers which don't extend much credit to one another.
+
+Suppose two peers, Alice and Bob, are operating the same or compatible settlement engine implementations, which settle using a payment channel on a shared ledger.
+
+1. Alice instructs her settlement engine to create a payment channel to Bob and fund it for 20 units.
+   - Alice's settlement engine should escrow 20 units in the payment channel and inform Bob's settlement engine that a payment channel was created.
+   - Note, however, that the funds are still in Alice's custody, and are still only accessible by Alice.
+2. Alice's accounting system instructs her settlement engine to settle 5 units to Bob.
+3. Alice's settlement engine signs a claim entitling Bob to 5 units from Alice's escrowed funds, and sends it to Bob's settlement engine.
+   - If Bob already had some claim, the new claim should entitle Bob to 5 _additional_ units.
+   - If Alice has already entitled Bob to all the available escrowed funds, the settlement engine may track that it owes Bob 5 units without sending a claim, until there's sufficient capacity in the payment channel.
+4. Bob's settlement engine, upon receiving the claim, should verify the message is valid and its digital signature is authentic, and determine how much additional funds it's entitled to (in this case, 5 units). Then, it should it notify the accounting system that it received the given amount, 5 units.
+5. Alice and Bob may repeat this process many times, as long as the party sending the payment channel claim has funds that haven't already been allocated to the peer. If additional funds need to be escrowed, the settlement engine may offer this functionality through a custom instruction.
+6. At a later time, Alice or Bob may instruct their settlement engines to close the payment channel and release the allotted funds back to their respective ledger accounts.
+   - Depending upon the payment channel implementation, there may be both cooperative and non-cooperative mechanisms for this, with varying time delays until the funds can be dispersed.
+   - Settlement engines using payment channels may also have functionality to watch the underlying ledger to ensure the peer doesn't try to submit an old message that would incorrectly distribute the escrowed funds.
+
+### Sender collects a fee
+
+Suppose Alice and Bob's settlement arrangement is such that, in order the Alice to settle her liabilities to Bob, Alice requires that first Bob must pay a fee.
```

Comment:
Why would you do this instead of just subtracting the fee amount from how much Alice is settling to Bob? aren't the two equivalent?

Reply:
Because it breaks the contract between the accounting system and SE: if Alice says "settle x," Bob--at some point--needs to say "receipt x." This causes their balances to get out-of-sync.

Here's a practical reason: Bob needs to fulfill enough ILP packets in order for Alice to say "settle" to her SE, so the fee will be marked as paid. _But_ Bob's accounting system doesn't know to fulfill enough packets to cover the fee! The idea is that the fee is entirely internal to the SE: you shouldn't have to manually config your max balance to budget for it. So, Bob's SE can just negotiate the fee with Alice, say "got money" for the value of the fee, _and then_ that allows enough packets to be fulfilled so Alice will say "settle" which marks the fee as paid -- and then allows the normal settlement flow to proceed afterwards.

Reply:
Is this a setup fee or a fee that applies to each settlement?

Reply:
Either, depending upon the implementation or use case.